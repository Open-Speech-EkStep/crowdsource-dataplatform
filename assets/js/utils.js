const { HOUR_IN_SECONDS, SIXTY } = require("./constants");

const ALL_LANGUAGES = [
    {value: "Assamese",id: "as", text: "অসমীয়া", hasLocaleText: true},
    {value: "Bengali", id: "bn", text: "বাংলা", hasLocaleText: true},
    {value: "English", id: "en", text: "English", hasLocaleText: true},
    {value: "Gujarati", id: "gu", text: "ગુજરાતી", hasLocaleText: true},
    {value: "Hindi", id: "hi", text: "हिंदी", hasLocaleText: true},
    {value: "Kannada", id: "kn", text: "ಕನ್ನಡ", hasLocaleText: true},
    {value: "Malayalam", id: "ml", text: "മലയാളം", hasLocaleText: true},
    {value: "Marathi", id: "mr", text: "मराठी", hasLocaleText: true},
    {value: "Odia", id: "or", text: "ଓଡିଆ", hasLocaleText: true},
    {value: "Punjabi", id: "pa", text: "ਪੰਜਾਬੀ", hasLocaleText: true},
    {value: "Tamil", id: "ta", text: "தமிழ்", hasLocaleText: true},
    {value: "Telugu", id: "te", text: "తెలుగు", hasLocaleText: true}];

function convertPXToVH(px) {
    return px * (100 / document.documentElement.clientHeight);
}

function setPageContentHeight() {
    const $footer = $('footer');
    const $nav = $('.navbar');
    const edgeHeightInPixel = $footer.outerHeight() + $nav.outerHeight()
    const contentHeightInVH = 100 - convertPXToVH(edgeHeightInPixel)
    $('#content-wrapper').css('min-height', contentHeightInVH + 'vh');
}

function toggleFooterPosition() {
    const $footer = $('footer');
    $footer.toggleClass('fixed-bottom')
    $footer.toggleClass('bottom')
}

function fetchLocationInfo() {
    //https://api.ipify.org/?format=json
    return fetch('https://www.cloudflare.com/cdn-cgi/trace').then(res => res.text()).then(ipAddressText => {
        const dataArray = ipAddressText.split('\n');
        let ipAddress = "";
        for (let ind in dataArray) {
            if (dataArray[ind].startsWith("ip=")) {
                ipAddress = dataArray[ind].replace('ip=', '');
                break;
            }
        }
        if (ipAddress.length !== 0) {
            return fetch(`/location-info?ip=${ipAddress}`);
        } else {
            return new Promise((resolve, reject) => {
                reject("Ip Address not available")
            })
        }
    });
}

const performAPIRequest = (url) => {
    return fetch(url).then((data) => {
        if(!data.ok) {
            throw Error(data.statusText || 'HTTP error');
        } else {
            return Promise.resolve(data.json());
        }
    });
}

const getLocaleString = function() {
    performAPIRequest('/get-locale-strings')
        .then((response) => {
            localStorage.setItem('localeString', JSON.stringify(response));
        });
}

const updateLocaleLanguagesDropdown = (language) => {
    const dropDown = $('#localisation_dropdown');
    const localeLang = ALL_LANGUAGES.find(ele => ele.value === language);
    if (language.toLowerCase() === "english" || localeLang.hasLocaleText === false) {
        dropDown.html('<a id="english" class="dropdown-item" href="/changeLocale/en">English</a>');
    } else {
        dropDown.html(`<a id="english" class="dropdown-item" href="/changeLocale/en">English</a>
        <a id=${localeLang.value} class="dropdown-item" href="/changeLocale/${localeLang.id}">${localeLang.text}</a>`);
    }
}

const calculateTime = function (totalSeconds, isSeconds = true) {
    const hours = Math.floor(totalSeconds / HOUR_IN_SECONDS);
    const remainingAfterHours = totalSeconds % HOUR_IN_SECONDS;
    const minutes = Math.floor(remainingAfterHours / SIXTY);
    const seconds = Math.round(remainingAfterHours % SIXTY);
    if (isSeconds) {
        return { hours, minutes, seconds };
    } else {
        return { hours, minutes };
    }
};

const formatTime = function (hours, minutes = 0, seconds = 0) {
    let result = "";
    if (hours > 0) {
        result += `${hours} hrs `;
    }
    if (minutes > 0) {
        result += `${minutes} min `;
    }
    if (hours === 0 && minutes === 0 && seconds > 0) {
        result += `${seconds} sec `;
    }
    return result.substr(0, result.length - 1);
};

module.exports = { setPageContentHeight, toggleFooterPosition, fetchLocationInfo, updateLocaleLanguagesDropdown ,calculateTime, formatTime, getLocaleString, performAPIRequest}
