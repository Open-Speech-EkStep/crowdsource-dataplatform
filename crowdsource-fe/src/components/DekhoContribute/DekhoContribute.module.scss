.root {
  z-index: 1;
}

.label {
  font-weight: 500;
}

.progress {
  height: 0.625rem;
}
