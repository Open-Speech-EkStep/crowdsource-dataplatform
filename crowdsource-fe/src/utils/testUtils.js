import { render } from '@testing-library/react';

const Wrapper = ({ children }) => children;

const customRender = (ui, options) => render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { renderHook } from '@testing-library/react-hooks';

// override render method
export { customRender as render };
