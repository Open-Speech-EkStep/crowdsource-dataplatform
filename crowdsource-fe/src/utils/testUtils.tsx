/* eslint-disable jest/no-export */
import type { ReactNode, ReactElement, ComponentType } from 'react';

import type { RenderOptions, RenderResult } from '@testing-library/react';
import { render, act } from '@testing-library/react';
import { axe } from 'jest-axe';
import puppeteer from 'puppeteer';
import type { Browser } from 'puppeteer';

const Wrapper: ComponentType = ({ children }: { children?: ReactNode }) => children as ReactElement;

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { renderHook } from '@testing-library/react-hooks';

// override render method
export { customRender as render };

export const verifyAxeTest = ({ container }: { container: RenderResult['container'] }) => {
  it('should not fail an axe audit', async () => {
    let axeResult;

    await act(async () => {
      axeResult = await axe(container);
    });

    expect(axeResult).toHaveNoViolations();
  });
};

export const verifyVRTest = (title: string, url: string) => {
  describe(`${title} VR test`, () => {
    let browser: Browser;

    beforeAll(async () => {
      browser = await puppeteer.launch({
        headless: true,
        devtools: false,
        ignoreHTTPSErrors: true,
        args: ['--start-fullscreen'],
      });
    });

    it('match snapshot', async () => {
      const page = await browser.newPage();
      await page.setViewport({ width: 1920, height: 1080 });

      await page.goto(url, {
        waitUntil: 'networkidle2',
      });

      await page.addStyleTag({
        content: `
      *,
      *::after,
      *::before {
          transition-delay: 0s !important;
          transition-duration: 0s !important;
          animation-delay: -0.0001s !important;
          animation-duration: 0s !important;
          animation-play-state: paused !important;
          caret-color: transparent !important;
      }
      `,
      });

      const image = await page.screenshot({
        fullPage: true,
      });

      expect(image).toMatchImageSnapshot();
    });

    afterAll(async () => {
      await browser.close();
    });
  });
};
