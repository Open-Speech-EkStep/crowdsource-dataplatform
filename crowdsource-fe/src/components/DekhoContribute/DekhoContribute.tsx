import { Fragment, useEffect, useState } from 'react';

import { useTranslation } from 'next-i18next';
import Image from 'next/image';
import { useRouter } from 'next/router';
import Container from 'react-bootstrap/Container';
import ProgressBar from 'react-bootstrap/ProgressBar';
import Spinner from 'react-bootstrap/Spinner';

import ButtonControls from 'components/ButtonControls';
import ChromeExtension from 'components/ChromeExtension';
import FunctionalHeader from 'components/FunctionalHeader';
import ImageView from 'components/ImageView';
import NoDataFound from 'components/NoDataFound';
import TextEditArea from 'components/TextEditArea';
import apiPaths from 'constants/apiPaths';
import {
  INITIATIVE_ACTIONS,
  INITIATIVES_MAPPING,
  INITIATIVES_MEDIA_MAPPING,
} from 'constants/initiativeConstants';
import { TEXT_INPUT_LENGTH } from 'constants/Keyboard';
import localStorageConstants from 'constants/localStorageConstants';
import routePaths from 'constants/routePaths';
import { useSubmit } from 'hooks/useFetch';
import useLocalStorage from 'hooks/useLocalStorage';
import useFetch from 'hooks/usePostFetch';
import type { ActionStoreInterface } from 'types/ActionRequestData';
import type { LocationInfo } from 'types/LocationInfo';
import type SpeakerDetails from 'types/SpeakerDetails';
import { getBrowserInfo, getDeviceInfo } from 'utils/utils';

import styles from './DekhoContribute.module.scss';

interface ResultType {
  data: any;
}

const DekhoContribute = () => {
  const { t } = useTranslation();

  const [contributionLanguage] = useLocalStorage<string>(localStorageConstants.contributionLanguage);

  const [speakerDetails] = useLocalStorage<SpeakerDetails>(localStorageConstants.speakerDetails);

  const [locationInfo] = useLocalStorage<LocationInfo>(localStorageConstants.locationInfo);

  const [showThankyouMessage, setShowThankyouMessage] = useState(false);

  const [closeKeyboard, setCloseKeyboard] = useState(false);

  const [contributionData, setContributionData] = useState([]);
  const [currentDataIndex, setCurrentDataIndex] = useState<number>(0);
  const [hasError, setHasError] = useState(false);
  const router = useRouter();
  const { locale: currentLocale } = useRouter();
  const [showUIData, setShowUIdata] = useState({
    media_data: '',
    dataset_row_id: '0',
  });

  const { submit } = useSubmit(apiPaths.store);

  const { submit: submitSkip } = useSubmit(apiPaths.skip);

  const [formData, setFormData] = useState<ActionStoreInterface>({
    userInput: '',
    speakerDetails: '',
    language: contributionLanguage ?? '',
    type: INITIATIVES_MEDIA_MAPPING.dekho,
    sentenceId: 0,
    state: '',
    country: '',
    device: getDeviceInfo(),
    browser: getBrowserInfo(),
  });

  const result = useFetch<ResultType>({
    url: apiPaths.mediaOCR,
    init: contributionLanguage
      ? {
          body: JSON.stringify({
            language: contributionLanguage,
            userName: speakerDetails?.userName,
          }),
          method: 'POST',
          credentials: 'include',
          mode: 'cors',
          headers: { 'Content-Type': 'application/json' },
        }
      : undefined,
  });

  useEffect(() => {
    if (result && result.data) {
      setContributionData(result.data);
      setShowUIdata(result.data[currentDataIndex]);
    }
  }, [currentDataIndex, result]);

  useEffect(() => {
    if (speakerDetails) {
      setCurrentDataIndex(0);
    }
  }, [speakerDetails]);

  const onChangeTextInput = (text: string) => {
    setFormData({
      ...formData,
      userInput: text,
    });
  };

  const setDataCurrentIndex = (index: number) => {
    if (index === contributionData.length - 1) {
      router.push(`/${currentLocale}${routePaths.dekhoIndiaContributeThankYou}`, undefined, {
        locale: currentLocale,
      });
    } else {
      setCurrentDataIndex(index + 1);
      setShowUIdata(contributionData[index + 1]);
    }
  };

  const resetState = () => {
    onCancelContribution();
  };

  const onSubmitContribution = () => {
    setShowThankyouMessage(true);
    setCloseKeyboard(!closeKeyboard);
    setDataCurrentIndex(currentDataIndex);
    resetState();
    submit(
      JSON.stringify({
        ...formData,
        language: contributionLanguage,
        sentenceId: showUIData.dataset_row_id,
        country: locationInfo?.country,
        state: locationInfo?.regionName,
        speakerDetails: JSON.stringify({
          userName: speakerDetails?.userName,
        }),
      })
    );
    setTimeout(() => {
      setShowThankyouMessage(false);
    }, 1500);
  };

  const onCancelContribution = () => {
    setFormData({
      ...formData,
      userInput: '',
    });
  };

  const onSkipContribution = () => {
    setDataCurrentIndex(currentDataIndex);
    setCloseKeyboard(!closeKeyboard);
    resetState();
    submitSkip(
      JSON.stringify({
        device: getDeviceInfo(),
        browser: getBrowserInfo(),
        userName: speakerDetails?.userName,
        language: contributionLanguage,
        sentenceId: showUIData.dataset_row_id,
        state_region: locationInfo?.regionName,
        country: locationInfo?.country,
        type: INITIATIVES_MEDIA_MAPPING.dekho,
      })
    );
  };

  if (!result) {
    return <Spinner data-testid="PageSpinner" animation="border" variant="light" />;
  }

  return contributionData && result?.data?.length !== 0 ? (
    <Fragment>
      <ChromeExtension />
      <div className="pt-4 px-2 px-lg-0 pb-8">
        <FunctionalHeader
          onSuccess={onSkipContribution}
          type="contribute"
          initiative={INITIATIVES_MAPPING.dekho}
          action={INITIATIVE_ACTIONS[INITIATIVES_MAPPING.dekho]['contribute']}
          showSpeaker={false}
        />
        <Container fluid="lg" className="mt-5">
          <div data-testid="DekhoContribute" className={`${styles.root}`}>
            <div className="align-items-center text-center">
              <span className="display-3">{t(`${INITIATIVES_MAPPING.dekho}ContributionHeading`)}</span>
            </div>
            <div className="mt-2 mt-md-4">
              <ImageView imageUrl={showUIData?.media_data} />
            </div>
            <div className="mt-4 mt-md-8">
              <TextEditArea
                id="addText"
                isTextareaDisabled={false}
                language={contributionLanguage ?? ''}
                initiative={INITIATIVES_MAPPING.dekho}
                setTextValue={onChangeTextInput}
                textValue={formData.userInput}
                label={`${t('addText')}${
                  contributionLanguage && ` (${t(contributionLanguage.toLowerCase())})`
                }`}
                onError={setHasError}
                closeKeyboard={closeKeyboard}
              />
            </div>
            {showThankyouMessage ? (
              <div className="d-flex align-items-center justify-content-center mt-9 display-1">
                <span className="me-2 d-flex">
                  <Image src="/images/check_mark.svg" width="40" height="40" alt="check" />
                </span>
                {t('thankyouForContributing')}
              </div>
            ) : (
              <div className="mt-9 mt-md-12">
                <ButtonControls
                  cancelDisable={!formData.userInput}
                  submitDisable={formData.userInput.length < TEXT_INPUT_LENGTH.LENGTH || hasError}
                  onSubmit={onSubmitContribution}
                  onCancel={onCancelContribution}
                  onSkip={onSkipContribution}
                  playButton={false}
                />
              </div>
            )}

            <div className="d-flex align-items-center mt-10 mt-md-14">
              <div className="flex-grow-1">
                <ProgressBar
                  now={(currentDataIndex + 1) * (100 / contributionData.length)}
                  variant="primary"
                  className={styles.progress}
                />
              </div>
              <span className="ms-5">
                {currentDataIndex + 1}/{contributionData.length}
              </span>
            </div>
          </div>
        </Container>
      </div>
    </Fragment>
  ) : (
    <div className="d-flex flex-grow-1 align-items-center">
      <NoDataFound
        url={routePaths.dekhoIndiaHome}
        title={t('ocrContributeNoDataThankYouMessage')}
        text={t('noDataMessage', { language: t(`${contributionLanguage?.toLowerCase()}`) })}
        buttonLabel={t('ocrBackToInitiativePrompt')}
      />
    </div>
  );
};

export default DekhoContribute;
