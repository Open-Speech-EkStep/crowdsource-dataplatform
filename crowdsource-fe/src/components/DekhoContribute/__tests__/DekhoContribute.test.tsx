import { when } from 'jest-when';

import { render, screen, userEvent, waitFor } from 'utils/testUtils';

import DekhoContribute from '../DekhoContribute';

describe('DekhoContribute', () => {
  const resultData = {
    data: [
      {
        dataset_row_id: 1248671,
        media_data: 'inbound/ocr/Hindi/test.png',
      },
      {
        dataset_row_id: 1248672,
        media_data: 'inbound/ocr/Hindi/test2.png',
      },
      {
        dataset_row_id: 1248673,
        media_data: 'inbound/ocr/Hindi/test3.png',
      },
      {
        dataset_row_id: 1248674,
        media_data: 'inbound/ocr/Hindi/test4.png',
      },
      {
        dataset_row_id: 1248675,
        media_data: 'inbound/ocr/Hindi/test5.png',
      },
    ],
  };
  const setup = async (resultData: any) => {
    when(localStorage.getItem)
      .calledWith('contributionLanguage')
      .mockImplementation(() => 'Hindi');

    const locationInfo = {
      country: 'Test Country',
      regionName: 'Sample Region',
    };

    when(localStorage.getItem)
      .calledWith('locationInfo')
      .mockImplementation(() => JSON.stringify(locationInfo));

    const speakerDetails = {
      userName: 'abc',
      motherTongue: '',
      age: '',
      gender: '',
      language: 'Hindi',
      toLanguage: '',
    };

    when(localStorage.getItem)
      .calledWith('speakerDetails')
      .mockImplementation(() => JSON.stringify(speakerDetails));

    fetchMock.doMockOnceIf('/media/ocr').mockResponseOnce(JSON.stringify(resultData));
    const renderResult = render(<DekhoContribute />);

    await waitFor(() => {
      expect(fetchMock).toBeCalledWith('/media/ocr', {
        body: JSON.stringify({
          language: 'Hindi',
          userName: 'abc',
        }),
        headers: { 'Content-Type': 'application/json' },
        method: 'POST',
        credentials: 'include',
        mode: 'cors',
      });
    });

    await waitFor(() => expect(screen.queryByTestId('PageSpinner')).not.toBeInTheDocument());

    return renderResult;
  };

  it('should render the component and matches it against stored snapshot', async () => {
    const { asFragment } = await setup(resultData);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should show the thank you message when no data present', async () => {
    const result = { data: [] };
    await setup(result);

    await waitFor(() => {
      expect(screen.getByText('ocrContributeNoDataThankYouMessage')).toBeInTheDocument();
    });
  });

  it('should show the error popup when api throw the error and close modal on clicking button', async () => {
    const url = '/media/ocr';
    const errorResponse = new Error('Some error');
    fetchMock.doMockOnceIf(url).mockRejectOnce(errorResponse);
    render(<DekhoContribute />);
    await waitFor(() => {
      expect(fetchMock).toBeCalledWith('/media/ocr', {
        body: JSON.stringify({
          language: 'Hindi',
          userName: 'abc',
        }),
        headers: { 'Content-Type': 'application/json' },
        method: 'POST',
        credentials: 'include',
        mode: 'cors',
      });
    });

    await waitFor(() => {
      expect(screen.getByText('apiFailureError')).toBeInTheDocument();
    });

    await waitFor(() => {
      userEvent.click(screen.getByRole('button', { name: 'proceed' }));
    });

    await waitFor(() => {
      expect(screen.queryByRole('button', { name: 'proceed' })).not.toBeInTheDocument();
    });
  });

  it('should test the textarea text with valid language', async () => {
    await setup(resultData);

    userEvent.type(screen.getByRole('textbox', { name: 'addText (hindi)' }), 'बपपप');

    await waitFor(() => {
      expect(screen.queryByText('Please type in your chosen language')).not.toBeInTheDocument();
    });
  });

  it('should test the cancel button functionality', async () => {
    await setup(resultData);

    expect(screen.getByRole('button', { name: 'cancel' })).toBeDisabled();

    userEvent.type(screen.getByRole('textbox', { name: 'addText (hindi)' }), 'बपपप');

    expect(screen.getByRole('button', { name: 'cancel' })).toBeEnabled();

    userEvent.click(screen.getByRole('button', { name: 'cancel' }));

    expect(screen.getByRole('button', { name: 'cancel' })).toBeDisabled();

    await waitFor(() => {
      expect(screen.getByRole('textbox', { name: 'addText (hindi)' })).toHaveValue('');
    });
  });

  it('should test the skip functionality', async () => {
    const url = '/skip';
    const successResponse = { message: 'Skipped successfully.', statusCode: 200 };

    await setup(resultData);
    fetchMock.doMockOnceIf(url).mockResponseOnce(JSON.stringify(successResponse));

    expect(screen.getByRole('button', { name: 'skip' })).toBeEnabled();

    userEvent.click(screen.getByRole('button', { name: 'skip' }));

    await waitFor(() => {
      expect(fetchMock).toBeCalledWith(url, {
        method: 'POST',
        credentials: 'include',
        mode: 'cors',
        headers: { 'Content-Type': 'application/json' },
        body:
          expect.stringContaining('"userName":"abc"') &&
          expect.stringContaining('"language":"Hindi"') &&
          expect.stringContaining('"sentenceId":"1248671"') &&
          expect.stringContaining('"type":"ocr"'),
      });
    });
  });

  it('should test the submit button', async () => {
    const url = '/store';
    const successResponse = { success: true };

    await setup(resultData);

    fetchMock.doMockOnceIf(url).mockResponseOnce(JSON.stringify(successResponse));

    expect(screen.getByRole('button', { name: 'submit' })).toBeDisabled();

    userEvent.type(screen.getByRole('textbox', { name: 'addText (hindi)' }), 'बपपप');

    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'submit' })).toBeEnabled();
    });

    userEvent.click(screen.getByRole('button', { name: 'submit' }));

    await waitFor(() => {
      expect(fetchMock).toBeCalledWith(url, {
        method: 'POST',
        credentials: 'include',
        mode: 'cors',
        headers: { 'Content-Type': 'application/json' },
        body:
          expect.stringContaining('"userInput":"बपपप"') &&
          expect.stringContaining('"language":"Hindi"') &&
          expect.stringContaining('"sentenceId":"1248671"') &&
          expect.stringContaining('"speakerDetails":{"userName":"abc"}') &&
          expect.stringContaining('"type":"ocr"'),
      });
    });

    await waitFor(() => expect(screen.queryByRole('img', { name: 'check' })).toBeInTheDocument());

    await waitFor(() => expect(screen.queryByRole('img', { name: 'check' })).not.toBeInTheDocument());
  });

  it('should go to thank you page after 5 skip sentences', async () => {
    await setup(resultData);

    expect(screen.getByRole('button', { name: 'skip' })).toBeEnabled();

    userEvent.click(screen.getByRole('button', { name: 'skip' }));

    userEvent.click(screen.getByRole('button', { name: 'skip' }));

    userEvent.click(screen.getByRole('button', { name: 'skip' }));

    userEvent.click(screen.getByRole('button', { name: 'skip' }));

    userEvent.click(screen.getByRole('button', { name: 'skip' }));
  });
});
