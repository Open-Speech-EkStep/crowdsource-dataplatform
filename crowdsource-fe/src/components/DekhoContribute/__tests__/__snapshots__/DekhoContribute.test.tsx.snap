// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DekhoContribute should render the component and matches it against stored snapshot 1`] = `
<DocumentFragment>
  <div
    class="root d-none d-md-block d-lg-flex flex-wrap position-relative align-items-md-center bg-secondary px-3 ps-md-6 pe-md-12 py-3 border-bottom border-1 border-primary-10"
    data-testid="ChromeExtension"
  >
    <span
      class="fst-italic me-5"
    >
      chromeExtensionLabel
    </span>
    <div
      class="d-flex align-items-center mt-md-2 mt-lg-0"
    >
      <a
        class="flex-shrink-0"
        href="https://chrome.google.com/webstore/detail/google-input-tools/mclkkofklkfljcocdinagocijmpgbhab"
        rel="noreferrer"
        target="_blank"
      >
        installNow
      </a>
      <span
        class="sep d-flex border border-start-0 border-primary w-0 mx-5"
      />
      <button
        class="ceBtn flex-shrink-0 light secondary btn btn-light"
        data-testid="Button"
        type="button"
      >
        watchTheVideo
      </button>
    </div>
    <button
      class="close d-flex position-absolute cursor info normal btn btn-link"
      data-testid="Button"
      type="button"
    >
      <div
        style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
      >
        <div
          style="box-sizing: border-box; display: block; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjAiIGhlaWdodD0iMjAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
            style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
          />
        </div>
        <img
          alt="Close Chrome Extension"
          data-nimg="intrinsic"
          decoding="async"
          src="/_next/image?url=%2Fimages%2Fclose.svg&w=48&q=75"
          srcset="/_next/image?url=%2Fimages%2Fclose.svg&w=32&q=75 1x, /_next/image?url=%2Fimages%2Fclose.svg&w=48&q=75 2x"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </div>
    </button>
  </div>
  <div
    class="pt-4 px-2 px-lg-0 pb-8"
  >
    <header>
      <div
        class="d-flex justify-content-between align-items-center px-3 px-md-6"
      >
        <div
          class="root d-flex align-items-center"
          data-testid="Breadcrumbs"
        >
          <a
            class="link"
            href="/dekho-india"
          >
            <div
              class="d-flex align-items-center cursor-pointer"
            >
              <div
                class="icon d-flex"
              >
                <div
                  style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
                >
                  <div
                    style="box-sizing: border-box; display: block; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDciIGhlaWdodD0iNDIiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                      style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                    />
                  </div>
                  <img
                    alt="dekhoLogo"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="/_next/image?url=%2Fimages%2Fen%2Flogos%2Fen-dekhoIndiaLogo.svg&w=96&q=75"
                    srcset="/_next/image?url=%2Fimages%2Fen%2Flogos%2Fen-dekhoIndiaLogo.svg&w=48&q=75 1x, /_next/image?url=%2Fimages%2Fen%2Flogos%2Fen-dekhoIndiaLogo.svg&w=96&q=75 2x"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </div>
              </div>
              <span
                class="initiative d-none d-md-block ms-1 ms-md-3 font-family-rowdies"
              >
                dekho india
              </span>
            </div>
          </a>
          <span
            class="mx-1 mx-md-3 text-primary-40"
          >
            /
          </span>
          <span
            class="display-5"
          >
            label
          </span>
        </div>
        <div
          class="d-flex"
        >
          <div>
            <button
              class="root rounded d-flex align-items-center px-3 text-primary text-decoration-none info normal btn btn-link"
              data-testid="reportIconAlt_button"
              type="button"
            >
              <div
                style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
              >
                <div
                  style="box-sizing: border-box; display: block; max-width: 100%;"
                >
                  <img
                    alt=""
                    aria-hidden="true"
                    src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                    style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                  />
                </div>
                <img
                  alt="reportIconAlt"
                  data-nimg="intrinsic"
                  decoding="async"
                  src="/_next/image?url=%2Fimages%2Freport.svg&w=48&q=75"
                  srcset="/_next/image?url=%2Fimages%2Freport.svg&w=32&q=75 1x, /_next/image?url=%2Fimages%2Freport.svg&w=48&q=75 2x"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
                <noscript />
              </div>
              <span
                class="text ms-2 d-none d-md-block"
              >
                report
              </span>
            </button>
          </div>
        </div>
      </div>
    </header>
    <div
      class="mt-5 container-lg"
    >
      <div
        class="root"
        data-testid="DekhoContribute"
      >
        <div
          class="mt-4 mt-md-8"
        >
          <div
            class="align-items-center text-center"
          >
            <span
              class="label display-3"
            >
              dekhoContributionHeading
            </span>
          </div>
          <div
            class="mt-9 mt-md-12"
          >
            <div
              class="d-flex flex-column"
            >
              <div
                class="bg bg-light p-8 rounded-8 border border-1 border-primary-20 position-relative"
              >
                <img
                  alt="OCR Data"
                  class="img-fluid"
                  src="/inbound/ocr/Hindi/test.png"
                />
              </div>
              <button
                class="display-6 ms-auto mt-1 text-primary-60 info normal btn btn-link"
                data-testid="Button"
                type="button"
              >
                expandViewText
              </button>
            </div>
          </div>
          <div
            class="mt-4 mt-md-8"
          >
            <div
              class="position-relative p-4 addText rounded-8"
              data-testid="TextEditArea"
            >
              <div
                class="false"
              >
                <div>
                  <label
                    class="textareaLabel display-6 form-label"
                    for="addText"
                  >
                    addText (hindi)
                  </label>
                  <textarea
                    class="textarea border-0 p-0 display-3 form-control"
                    id="addText"
                    placeholder="typingPlaceholder"
                  />
                </div>
                <div
                  class="keyboardIcon d-none  d-md-block  position-absolute"
                >
                  <button
                    class="d-flex info normal btn btn-link"
                    data-testid="Button"
                    type="button"
                  >
                    <div
                      style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
                    >
                      <div
                        style="box-sizing: border-box; display: block; max-width: 100%;"
                      >
                        <img
                          alt=""
                          aria-hidden="true"
                          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                          style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                        />
                      </div>
                      <img
                        alt="keyboardBtn"
                        data-nimg="intrinsic"
                        decoding="async"
                        src="/_next/image?url=%2Fimages%2Fkeyboard_icon.svg&w=48&q=75"
                        srcset="/_next/image?url=%2Fimages%2Fkeyboard_icon.svg&w=32&q=75 1x, /_next/image?url=%2Fimages%2Fkeyboard_icon.svg&w=48&q=75 2x"
                        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                      />
                      <noscript />
                    </div>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="mt-9 mt-md-12"
        >
          <div
            class="d-flex flex-column flex-md-row justify-content-md-center align-items-center"
          >
            <button
              class="mx-md-6 order-2 order-md-1 my-2 my-md-0 light secondary btn btn-light"
              data-testid="Button"
              disabled=""
              type="button"
            >
              cancel
            </button>
            <button
              class="mx-md-6 order-3 order-md-3 my-2 my-md-0 info primary btn btn-primary"
              data-testid="Button"
              disabled=""
              type="button"
            >
              submit
            </button>
          </div>
          <div
            class="d-flex justify-content-center mt-4 mt-md-13"
          >
            <button
              class="transparent tertiary btn btn-transparent"
              data-testid="Button"
              type="button"
            >
              skip
            </button>
          </div>
        </div>
        <div
          class="d-flex align-items-center mt-10 mt-md-14"
        >
          <div
            class="flex-grow-1"
          >
            <div
              class="progress progress"
            >
              <div
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="20"
                class="progress-bar bg-primary"
                role="progressbar"
                style="width: 20%;"
              />
            </div>
          </div>
          <span
            class="ms-5"
          >
            1/5
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
